api/dashboard/create.go

```diff
@@ -0,0 +1,186 @@
+//   '400':
+//     description: Bad request when creating dashboard
+//     schema:
+//       "$ref": "#/definitions/Error"
+//   '401':
+//     description: Unauthorized to create dashboard
+//     schema:
+//       "$ref": "#/definitions/Error"
+//   '500':
+//     description: Server error when creating dashboard
+//     schema:
+//       "$ref": "#/definitions/Error"
+
+// CreateDashboard represents the API handler to
+// create a dashboard in the configured backend.
+func CreateDashboard(c *gin.Context) {
+	// capture middleware values
+	u := user.Retrieve(c)
+
+	// capture body from API request
+	input := new(types.Dashboard)
+
+	err := c.Bind(input)
+	if err != nil {
+		retErr := fmt.Errorf("unable to decode JSON for new dashboard: %w", err)
+
+		util.HandleError(c, http.StatusBadRequest, retErr)
+
+		return
+	}
+
+	// ensure dashboard name is defined
+	if input.GetName() == "" {
+		util.HandleError(c, http.StatusBadRequest, fmt.Errorf("dashboard name must be set"))
+
+		return
+	}
+
+	// update engine logger with API metadata
+	//
+	// https://pkg.go.dev/github.com/sirupsen/logrus?tab=doc#Entry.WithFields
+	logrus.WithFields(logrus.Fields{
+		"user": u.GetName(),
+	}).Infof("creating new dashboard %s", input.GetName())
+
+	d := new(types.Dashboard)
+
+	// update fields in dashboard object
+	d.SetCreatedBy(u.GetName())
+	d.SetName(input.GetName())
+	d.SetCreatedAt(time.Now().UTC().Unix())
+	d.SetUpdatedAt(time.Now().UTC().Unix())
+	d.SetUpdatedBy(u.GetName())
+
+	// validate admins to ensure they are all active users
+	admins, err := validateAdminSet(c, u, input.GetAdmins())
+	if err != nil {
+		util.HandleError(c, http.StatusBadRequest, err)
+
+		return
+	}
+
+	d.SetAdmins(admins)
+
+	// validate repos to ensure they are all enabled
+	err = validateRepoSet(c, input.GetRepos())
+	if err != nil {
+		util.HandleError(c, http.StatusBadRequest, err)
+
+		return
+	}
+
+	d.SetRepos(input.GetRepos())
+
+	// create dashboard in database
+	d, err = database.FromContext(c).CreateDashboard(c, d)
+	if err != nil {
+		retErr := fmt.Errorf("unable to create new dashboard %s: %w", d.GetName(), err)
+
+		util.HandleError(c, http.StatusInternalServerError, retErr)
+
+		return
+	}
+
+	// add dashboard to claims' user's dashboard set
+	u.SetDashboards(append(u.GetDashboards(), d.GetID()))
+
+	// update user in database
+	_, err = database.FromContext(c).UpdateUser(c, u)
+	if err != nil {
+		retErr := fmt.Errorf("unable to update user %s: %w", u.GetName(), err)
+
+		util.HandleError(c, http.StatusInternalServerError, retErr)
+
+		return
+	}
+
+	c.JSON(http.StatusCreated, d)
+}
+
+// validateAdminSet takes a slice of user names and converts it into a slice of matching
+// user ids in order to preserve data integrity in case of name change.
+func validateAdminSet(c context.Context, caller *types.User, users []string) ([]string, error) {
+	// add user creating the dashboard to admin list
+	admins := []string{fmt.Sprintf("%d", caller.GetID())}
```

Comment:
should we move this create-specific logic to outside of this func?

Reply:
just to right before it gets called, wherever that is, its just hard to tell from a glance that validating the list of admins also adds an admin to the list. 
purely an opinion tho

Reply:
Oh I misunderstood. Yeah I agree.. but I'll just rename the function instead of move the logic outside